jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -4060,6 +4066,9 @@ static CacheKey newDeclaredPublicMethodsKey(String methodName, Class<?>[] parame
 		return new CacheKey("#m" + methodName, parameterTypes, null);	//$NON-NLS-1$
 	}
 
+	static final CacheKey ClassCanonicalNameKey = new CacheKey("CanonicalName", EmptyParameters, null); //$NON-NLS-1$
+	static final CacheKey ClassSimpleNameKey = new CacheKey("SimpleName", EmptyParameters, null); //$NON-NLS-1$
```

Comment:
@pshipton are you familiar with the rationale behind the naming scheme for CacheKey's using the `.X` or `/X` format?  These names seem fine to me but I'm not sure if they fit the existing schema.

Reply:
The names shouldn't conflict with names of any possible methods, otherwise the method and cached name will conflict in the cache. i.e. the current names do conflict. The special characters are used to avoid conflicts. The different characters distinguish between different types of reflect data. Any special character can be used to avoid conflicts, as long as the chosen name isn't already used in the cache.

The cache uses SoftReference to store the values, so the cached names will disappear after some time if not referenced. This seems like desirable behavior though, just pointing it out.

This change could result in a significant footprint increase due to creating a ConcurrentHashMap to hold the cached values. Many classes may not be accessed via reflection, but have their simple or canonical name accessed. The ConcurrentHashMap will be null'ed once the cached values are collected.

Reply:
With regards to the `CacheKey` I can add just a special symbol at the beginning to make it unique.

As for the memory footprint I think it makes sense. `ConcurrentHashMap` is pretty heavy actually. But I can suggest using something more lightweight, like `HashMap`. In this case multiple threads can set the same value but I don't think this is a big problem.

As an alternative, these values can be stored lazily (on demand) in a couple of `SoftReference`s.

Reply:
I had proposed reusing the reflect cache to avoid growing the default footprint of j.l.Class but as pointed out, the CHM is a lot of overhead here.  Looking at the j.l.Class, there are a number of existing caches:
```
* private transient AnnotationCache annotationCache;
* private transient AnnotationVars annotationVars;
* private transient ClassRepositoryHolder classRepoHolder;
* private transient ReflectCache reflectCache;
* private transient Class<?> cachedDeclaringClass;
* private transient Class<?> cachedEnclosingClass;  // getting enclosingClass also caches declaringClass
* private transient Class<?>[] cachedInterfaces;
* private transient Object methodHandleCache;
```

An alternative would be to add a new cache that holds the {cachedDeclaringClass, cachedEnclosingClass, cachedInterfaces} & the new two SoftReferences for cached names.

This would require more performance and footprint results to validate it doesn't regress startup / throughput / footprint.  

Thoughts?

Reply:
I did a draft implementation of this proposal. It would be nice if it can be checked.
And I have also updated the description with new benchmark results.